File: mobile_app/lib/user/models/teacher.dart
import 'package:mobile_app/shared/required_enums.dart';
import 'package:mobile_app/user/models/app_user.dart';

class Teacher extends AppUser {
  final String? avatarPath;
  final String? avatarURL;

  final String empId;
  final List<String> subjectIds;
  final List<String> classIds;
  final DateTime lastCheckedNotices;

  Teacher({
    required super.id,
    required super.name,
    required super.email,
    required super.createdAt,
    required super.role,
    this.avatarPath,
    this.avatarURL,
    required this.empId,
    required this.subjectIds,
    required this.classIds,
    required this.lastCheckedNotices,
  });

  factory Teacher.fromMap(
    Map<String, dynamic> user,
    Map<String, dynamic> teacherData,
  ) {
    return Teacher(
      id: user['id'],
      name: user['full_name'],
      email: user['email'],
      createdAt: DateTime.parse(user['created_at']).toLocal(),
      role: UserRoles.teacher,
      avatarPath: teacherData['avatar_path'],
      avatarURL: teacherData['avatar_url'],
      empId: teacherData['employee_id'],
      subjectIds: List<String>.from(teacherData['subject_ids']),
      classIds: List<String>.from(teacherData['class_ids']),
      lastCheckedNotices: DateTime.parse(teacherData['last_checked_notices']).toLocal(),
    );
  }
}

File: admin_web/src/models/teacher.ts
export class Teacher {
  id: string;
  fullName: string;
  email: string;
  employeeId: string;
  subjectIds: string[];
  classIds: string[];
  avatarPath: string | null;

  constructor(
    id: string,
    fullName: string,
    email: string,
    employeeId: string,
    subjectIds: string[],
    classIds: string[],
    avatarPath: string | null
  ) {
    this.id = id;
    this.fullName = fullName;
    this.email = email;
    this.employeeId = employeeId;
    this.subjectIds = subjectIds;
    this.classIds = classIds;
    this.avatarPath = avatarPath;
  }

  static fromMap(map: any): Teacher {
    return new Teacher(
      map.id,
      map.full_name || map.fullName || "",  // from users JOIN or memory
      map.email || "",                       // from users JOIN or memory
      map.employee_id || "",
      map.subject_ids || [],
      map.class_ids || [],
      map.avatar_path || null
    );
  }

  toMap() {
    return {
      id: this.id,
      // full_name and email live in users table — NOT in teacher table
      employee_id: this.employeeId,
      subject_ids: this.subjectIds,
      class_ids: this.classIds,
      avatar_path: this.avatarPath,
    };
  }

  toInsertMap() {
    return {
      id: this.id,
      // full_name and email live in users table — NOT in teacher table
      employee_id: this.employeeId,
      subject_ids: this.subjectIds,
      class_ids: this.classIds,
      avatar_path: this.avatarPath,
    };
  }
}